http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala

```diff
@@ -102,30 +89,40 @@ class SamAuthProvider(val config: Config, serviceAccountProvider: ServiceAccount
     */
-  override def hasNotebookClusterPermission(internalId: ClusterInternalId, userInfo: UserInfo, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+  override def hasNotebookClusterPermission(internalId: ClusterInternalId, userInfo: UserInfo, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Boolean] = {
+    val authorization = Authorization(Credentials.Token(AuthScheme.Bearer, userInfo.accessToken.token))
     // Consult the notebook auth cache if enabled
-    if (notebookAuthCacheEnabled) {
+    if (config.notebookAuthCacheEnabled) {
       // tokenExpiresIn should not taken into account when comparing cache keys
-      notebookAuthCache.get(NotebookAuthCacheKey(internalId, userInfo.copy(tokenExpiresIn = 0), action, googleProject, clusterName, executionContext))
+      Effect[F].delay(notebookAuthCache.get(NotebookAuthCacheKey(internalId, authorization, action, googleProject, clusterName)).booleanValue())
     } else {
-      hasNotebookClusterPermissionInternal(internalId, userInfo, action, googleProject, clusterName)
+      hasNotebookClusterPermissionInternal(internalId, authorization, action, googleProject, clusterName)
     }
   }
 
-  private def hasNotebookClusterPermissionInternal(internalId: ClusterInternalId, userInfo: UserInfo, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit executionContext: ExecutionContext): Future[Boolean] = {
-    // if action is connect, check only cluster resource. If action is anything else, either cluster or project must be true
-    retryUntilSuccessOrTimeout(shouldInvalidateSamCacheAndRetry, s"SamAuthProvider.hasNotebookClusterPermissionInternal call failed for ${googleProject.value}/${clusterName.value}")(samRetryInterval, samRetryTimeout) { () =>
-      Future {
-        val hasNotebookAction = blocking(samClient.hasActionOnNotebookClusterResource(internalId, userInfo, getNotebookActionString(action)))
-        if (action == ConnectToCluster) {
-          hasNotebookAction
-        } else {
-          hasNotebookAction || blocking(samClient.hasActionOnBillingProjectResource(userInfo, googleProject, getProjectActionString(action)))
-        }
+  private def hasNotebookClusterPermissionInternal(internalId: ClusterInternalId, authorization: Authorization, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Boolean] = {
+    for {
+      traceId <- ev.ask
+      hasNotebookAction <- hasNotebookClusterPermissionInternal(internalId, action, authorization)
+      res <- if(action == ConnectToCluster) Sync[F].pure(hasNotebookAction) else {
```

Comment:
I think this logic is correct, but it's a bit confusing so would be good to retain the above comment.
```
// if action is connect, check only cluster resource. If action is anything else, either cluster or project must be true
```